File: Study Plan/Top Interview 150/0205 Isomorphic Strings/IsomorphicStrings.cs
/*  
    Problem
    --------------------------------------------------------------------------------
    205. Isomorphic Strings
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    https://leetcode.com/problems/isomorphic-strings/solutions/4484437/c-time-and-space-complexity-o-n-detailed-step-by-step-explanation-dictionary-and-hashset/

    - Runtime 63 ms Beats (84.64% of users with C#)
    - Memory 42.33 MB (Beats 8.05% of users with C#)
*/

public class Solution
{
    public bool IsIsomorphic(string s, string t)
    {
        // Constraints: t.length == s.length
        var length = s.Length;

        // Map of source (Key) and target (Value) characters
        var charMap = new Dictionary<char, char>(length);
        // Set of characters that have already been mapped
        var charSet = new HashSet<char>(length);

        for (var i = 0; i < length; i++)
        {
            // Loop through the characters of both strings
            var source = s[i];
            var target = t[i];

            // Check if the character from the source has already been mapped
            if (charMap.ContainsKey(source))
            {
                // Verify that the mapping matches
                if (charMap[source] != target)
                {
                    // Inconsistency
                    return false;
                }
            }
            // The character hasn't been mapped yet
            else
            {
                // Check if the character from the target has already been mapped
                if (charSet.Contains(target))
                {
                    // Already mapped to another character
                    return false;
                }

                // Add the source/target character mapping
                charMap[source] = target;
                // Add the target character to the set of already mapped
                charSet.Add(target);
            }
        }

        return true;
    }
}

File: Study Plan/Top Interview 150/0205 Isomorphic Strings/IsomorphicStrings.java
/*  
    Problem
    --------------------------------------------------------------------------------
    205. Isomorphic Strings
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 10 ms (Beats 69.40% of users with Java)
    - Memory 43.11 MB (Beats 10.47% of users with Java)
*/

class Solution {
    public boolean isIsomorphic(String s, String t) {
        // Constraints: t.length == s.length
        int length = s.length();
        // map of source (Key) and target (Value) characters
        HashMap<Character, Character> charMap = new HashMap<>(length);
        // set of characters that have already been mapped
        HashSet<Character> charSet = new HashSet<>(length);
        for (int i = 0; i < length; i++) {
            // loop through the characters of both strings
            char source = s.charAt(i);
            char target = t.charAt(i);
            // check if the character from the source has already been mapped
            if (charMap.containsKey(source)) {
                // verify that the mapping matches
                if (charMap.get(source) != target) {
                    // inconsistency
                    return false;
                }
            // the character hasn't been mapped yet
            } else {
                // check if the character from the target has already been mapped
                if (charSet.contains(target)) {
                    // already mapped to another character
                    return false;
                }
                // add the source/target character mapping
                charMap.put(source, target);
                // add the target character to the set of already mapped
                charSet.add(target);
            }
        }
        return true;
    }
}